[14.0][IMP] account_asset_management, allow manual entry/editing of depreciation board

bosd opened the pull request:
Forward port of https://github.com/OCA/account-financial-tools/pull/1208 continuation of #1237

JordiBForgeFlow committed:
[IMP] account_asset_management: allow manual entry/editing of depreciation board.
- Enter manually the depreciation board.
- Compute the depreciation board based on rules, but allow editing of lines.

Some companies may need to manage a custom depreciation board because the depreciation schedule is too custom,
or because it may have evolved over time and cannot follow a single method.
In those cases it's better to manage a manual depreciation schedule.

[account_asset_management][fix] allow updating the previous_id

[account_asset_management][fix] comparison of dates

kittiu committed:
[IMP] account_asset_management: improve tests,  manual depreciation constraints

bosd commented:
@Dranyel-Bosd Can you please review?

Dranyel-Bosd reviewed (approved):
LGTM :+1: :sparkler:

bofiltd reviewed (approved):
LGTM

bosd commented:
@etobella Can you please merge?

etobella commented on account_asset_management/models/account_asset.py:
```diff
@@ -259,6 +259,14 @@ class AccountAsset(models.Model):
         states=READONLY_STATES,
         check_company=True,
     )
+    depreciation_line_manual_ids = fields.One2many(
```
Why do you need this field if you already have depreciation_line_ids?

bosd replied:
Great spot, maybe that one can be removed. I'll check

bosd replied:
Removed the field :+1:

bosd committed:
[IMP] account_asset_management: simplify,refactor manual depreciation

etobella commented on account_asset_management/models/account_asset.py:
```diff
@@ -279,6 +279,8 @@ class AccountAsset(models.Model):
         readonly=False,
         store=True,
     )
+    editable_lines = fields.Boolean(string="Allow editing the depreciation board")
```
I assume, that the first one only allows edition and the second one is for adding lines, isn't it?

bosd replied:
The first one, is to completely bypass the automitic/computed method so one could use a custom board.

bosd commented:
Improved coverage
@etobella it is ready now

etobella reviewed (commented):
Technically LGTM, however I will ask comments from @JordiBForgeFlow as original author of the change :wink:

bosd commented:
ping @JordiBForgeFlow Can you have a look here / merge? 🙏

etobella commented:
/ocabot merge patch

etobella commented:
Let's proceed with merge in order to avoid delay (we have given enough time)

pedrobaeza reviewed (changes requested):
Sorry for going late to the party, but I don't think we should merge as is.

- The number of commits is not acceptable.
- The implementation with so many new fields is not good.

I'm aborting the merge and check this deeply proposing the implementation today.

luc-demeyer reviewed (commented):
I am not sure if we really need this.
The depreciation table can already be edited before confirming an asset.
I also think we'll hit performance probelems for companies with a large niumber of assets by changing the store=true on a couple of fields (my largest customer has close to 100000 assets, but ecosoft has a customer with more)
@Saran440 what do you think ?

bosd commented:
> The depreciation table can already be edited before confirming an asset.

But we cannot edit a running asset. Which is quite common to do in some countries.
So, I think we need this.

@JordiBForgeFlow @kittiu What do you think?

JordiBForgeFlow commented:
@bosd I also think that we need this.

bosd commented:
> I'm aborting the merge and check this deeply proposing the implementation today.

@pedrobaeza Which changes would you like to see?

kittiu commented:
@bosd IMHO I think the current one that edit before submit is good enough. I think I am too concern with performance when dealing with lots of asset. Will this store false account into it?

If not, I think this can be a good feature.

bosd commented:
Update: Changed the fields back to stored.



> * The number of commits is not acceptable.

I have squashed the commits of @JordiBForgeFlow .
The other commits I kept in place to retain author ship.

Renamed the commit messages.

bosd commented:
@pedrobaeza Is this one ok now?

pedrobaeza commented:
I'm trying this, and it's very confusing to put manual depreciation board, add a line, and the error "You should at least have a starting depreciation line of Type 'Depreciation Base'." to appear, getting to a blocked situation, as that line can't be put manually nor computed.

Another problem is the `previous_id` computation, using compute hacks + double writing. Please change the approach. One possibility is to add the assets to "recompute its children", and then write all the `previous_id`.

Saran440 reviewed (commented):
@bosd Functional Test.

When we use Allow editing the depreciation board, Days in depre will not calculated.

![Selection_001](https://github.com/user-attachments/assets/07b98a83-b66a-459a-b73d-53cd67f24a6b)

bosd commented:
> @bosd Functional Test.
> 
> When we use Allow editing the depreciation board, Days in depre will not calculated.
> 

Imo that is a design intent. As when one does a manual depreciation it is not distributed evenly among the terms. 
@kittiu is that how you originally intended this?

bosd commented:
Can we prevent to loose this?

pedrobaeza commented:
My comments are not attended.

bosd commented:
Oopsie, Missed those..
Will come back to this...
